src/get-started/first-java-service.md

```diff
@@ -0,0 +1,615 @@
+    2. "Mark Directory As" > "Generated Sources Root"
+
+Compile the project:
+
+```shell
+mvn compile
+```
+
+<!--
+todo: Shall we add a whole file of MySchema?
+-->
+
+<!--
+Todo: Shall we draw a parallel with the DAO objects in business
+applications?
+-->
+
+<!--
+Todo: Shall we include the references to extra info on MerkleDB,
+indexes, serialization and serializers?
+-->
+
+### 3 Service Transactions
+
+Our service needs two operations updating its state:
+
+- Add a new vehicle entry to the registry
+- Transfer the ownership over the vehicle to another user.
+
+Modifying operations in Exonum are called transactions. Transactions are
+implemented as methods in a service class — a class implementing
+[`Service`][service-interface] interface. A transaction method must be annotated
+with [`@Transaction`][transaction-annotation] annotation.
+
+[service-interface]: https://exonum.com/doc/api/java-binding/0.10.0/com/exonum/binding/core/service/Service.html
+[transaction-annotation]: https://exonum.com/doc/api/java-binding/0.10.0/com/exonum/binding/core/transaction/Transaction.html
+
+Our project already has a template service `MyService` — navigate to it.
+
+#### Add Vehicle Transaction
+
+First, let's define the transaction arguments. As Exonum expects transaction
+arguments in a serialized form, we will define the arguments as a protobuf
+message.
+
+Add a new file `transactions.proto` in `car-registry-messages`
+(`car-registry-messages/src/main/proto/example/vehicle`).
+
+<!--
+todo: Again, shall we include a code example of an empty file
+(with package, option java_package, etc.)?
+-->
+
+Then add the message definition of the _Add Vehicle_ transaction:
+
+<!--codeinclude-->
+[](../../code-examples/java/exonum-java-binding/tutorials/car-registry/car-registry-messages/src/main/proto/example/vehicle/transactions.proto)
+inside_block:ci-add-vehicle
+<!--/codeinclude-->
+
+Compile the message:
+
+```shell
+mvn generate-sources
+```
+
+---
+
+Next, let's write the transaction method. Navigate back to `MyService`,
+and add the following method:
+
+<!--codeinclude-->
+[](../../code-examples/java/exonum-java-binding/tutorials/car-registry/car-registry-service/src/main/java/com/example/car/MyService.java)
+inside_block:ci-add-vehicle
+<!--/codeinclude-->
+
+The annotation accepts an integral _transaction ID_: an identifier of
+the transaction that service clients use to invoke it.
+
+The method accepts `AddVehicle` class: an auto-generated class from our
+protobuf definition. It also accepts the `TransactionContext` allowing
+the transaction to access the database.
+
+The transaction implementation uses the service data schema `MySchema`
+to access its data.
+
+Note that the transaction throws an `ExecutionException` in case of
+a precondition failure: in this case, an attempt to add a vehicle with
+an existing ID.
+
+<!--
+todo: Shall we add the ids to the example? If so, how do we split ids for
```

Comment:
What do you mean by "split"? IMO, the way transactions are defined in exonum/exonum-java-binding#1467 currently looks good.

Reply:
I mean that currently the snippets on the page do not include the ids, but symbolic references:

```java
@Transaction(FOO_TX_ID)
```

We can include them:


```java
private static int FOO_TX_ID = 0;

...
@Transaction(FOO_TX_ID)
```

but that might require grouping (splitting) the constants by transactions and not by their type (tx id or error code):

```
// ci-foo-tx {
foo_tx_id
error_code_from_foo
// }
// ci-bar-tx {
bar_tx_id
error_code_from_bar
// }
```

which I am not sure we shall do)

Reply:
I think it's okay to leave it as it is. The user would be able to lookup a transaction id if he wants.